File: report2/a2.py
import numpy as np

s_L = 4
e_L = 16


def get_e1_e2(filepath, threshold=1e-1):
    with open(filepath, 'r') as f:
        lines = f.readlines()
    line_number = next(i for i, line in enumerate(lines) if "Lanczos EigenValue in vec" in line)
    line_above = lines[line_number - 1]
    eigvals = line_above.split('=')[1].split(" ")[2:]
    eigvals = [np.float64(val) for val in eigvals]
    
    e1 = eigvals[0]
    i = 1
    while i < len(eigvals) and abs(e1 - eigvals[i]) < threshold:
        i += 1
    if i == len(eigvals):
        print("Did not find an eigenvalue distinct from e1 within threshold")
        return e1, None
    else:
        e2 = eigvals[i]
        if i != 1:
            print("Degenerated")
        return e1, e2


e1 = []
e2 = []
for L in range(s_L, e_L + 2, 2):
    print('L{}.result checked!'.format(L))
    e1_, e2_ = get_e1_e2("./L{}.result".format(L))
    e1.append(e1_)
    e2.append(e2_)

e1 = np.float64(e1)
e2 = np.float64(e2)

with open("e1.txt", 'w') as f:
    for e1_ in e1:
        f.write(str(e1_))
        f.write('\n')
with open("e2.txt", 'w') as f:
    for e2_ in e2:
        f.write(str(e2_))
        f.write('\n')

from scipy.optimize import curve_fit
import matplotlib.pyplot as plt

def func1(L, a, b, c):
    return a + b * np.exp(-c * L)

def func2(L, a, b, c):
    return a + b / L + c / (L * L)

x = np.float64(range(s_L, e_L + 2, 2))
popt, pcov = curve_fit(func2, x, e2 - e1)
print(popt)

plt.plot(x, func2(x, *popt), 'r-',
         label='fit: a=%5.3f, b=%5.3f, c=%5.3f' % tuple(popt))

plt.scatter(x, e2 - e1)

plt.savefig('a2.jpg')

File: report2/a2.jl
using DelimitedFiles
using LsqFit
using Plots
using LaTeXStrings
using Printf
Plots.pythonplot()
default(fontfamily="Monaco")
e1 = readdlm("e1.txt", Float64)
e2 = readdlm("e2.txt", Float64)
data = e2 - e1


x = collect(4:2:16) 
y = vec(data)
model(x, p) = p[1] .+ p[2] ./ x .+ p[3] ./ (x .^ 2) 


p0 = [rand(), rand(), rand(), rand(), rand(), rand(), rand()]

fit = curve_fit(model, x, y, p0)

title_str = @sprintf("\$c + \\frac{b}{L} + \\frac{a}{L^2}\$, c = %.2f, b = %.2f, a = %.2f", fit.param[1], fit.param[2], fit.param[3])
title_str = LaTeXString(title_str)

scatter(x, y, label="gap", xticks=4:2:16, title=title_str)


p = plot!(collect(4:2:32), model(collect(4:2:32), fit.param), label=L"c+\frac{b}{L}+\frac{a}{L^2}")


savefig(p, "a2.png")
